modules/library/metadata/src/main/java/org/geotools/factory/FactoryRegistry.java

```diff
@@ -152,15 +158,52 @@ public FactoryRegistry(final Class<?>[] categories) {
      * @param categories The categories.
      */
     public FactoryRegistry(final Collection<Class<?>> categories) {
-        super(categories.iterator());
-        for (final Iterator<Class<?>> it=getCategories(); it.hasNext();) {
-            if (needScanForPlugins == null) {
-                needScanForPlugins = new HashSet<Class<?>>();
-            }
-            needScanForPlugins.add(it.next());
+        for (Class<?> category : categories) {
+            needScanForPlugins.add(category);
+            registry.registerCategory(category);
         }
     }
 
+    /*
+     * TODO: discuss how much API can be broken
+     *
+     * Breakage:
+     *
+     *  - not all methods of `ServiceRegistry` API were implemented
+     *
+     * Possible breakage:
+     *
+     *  - in general nomenclature, replace "ServiceProvider" with "Factory"
+     *  - RegisterableService interface can no longer be used because it expects a `ServiceProvider`
+     *  - replace `Filter` with `Predicate<? super T>`
```

Comment:
I don't see a way to get rid of the generics by subtyping. There is no sensible `T` from which all factories extend (or did I miss it?). The `? super T` _could_ be replaced by `T` but that would make using the API more painful.

Should I do this together with renaming the methods? Then the new methods with their shiny new names take predicates and the old ones accept filter?